File: ios/SsdidWallet/Domain/Verifier/Offline/VerificationOrchestrator.swift
import Foundation

/// Provides bundle TTL configuration for the orchestrator.
protocol TtlProvider {
    var bundleTtlSeconds: TimeInterval { get }
}

/// Default TTL provider: 7-day bundle freshness window.
struct DefaultTtlProvider: TtlProvider {
    let bundleTtlSeconds: TimeInterval = 7 * 86400
}

/// Orchestrates credential verification by attempting online first,
/// then falling back to offline (cached bundle) verification when
/// network errors or server failures occur.
final class VerificationOrchestrator {
    private let onlineVerifier: Verifier
    private let offlineVerifier: OfflineVerifier
    private let bundleStore: BundleStore
    private let ttlProvider: TtlProvider

    init(
        onlineVerifier: Verifier,
        offlineVerifier: OfflineVerifier,
        bundleStore: BundleStore,
        ttlProvider: TtlProvider = DefaultTtlProvider()
    ) {
        self.onlineVerifier = onlineVerifier
        self.offlineVerifier = offlineVerifier
        self.bundleStore = bundleStore
        self.ttlProvider = ttlProvider
    }

    /// Verify a credential online, falling back to offline if a network/server error occurs.
    func verify(credential: VerifiableCredential) async -> UnifiedVerificationResult {
        do {
            let verified = try await onlineVerifier.verifyCredential(credential: credential)
            if verified {
                return UnifiedVerificationResult(
                    status: .verified,
                    checks: [
                        VerificationCheck(type: .signature, status: .pass, message: "Signature verified online"),
                        VerificationCheck(type: .expiry, status: .pass, message: "Credential is not expired"),
                        VerificationCheck(type: .revocation, status: .pass, message: "Revocation status confirmed online")
                    ],
                    source: .online
                )
            } else {
                return UnifiedVerificationResult(
                    status: .failed,
                    checks: [
                        VerificationCheck(type: .signature, status: .fail, message: "Signature verification failed")
                    ],
                    source: .online
                )
            }
        } catch {
            if isNetworkError(error) {
                return await fallbackToOffline(credential: credential)
            } else {
                return UnifiedVerificationResult(
                    status: .failed,
                    checks: [
                        VerificationCheck(type: .signature, status: .fail, message: error.localizedDescription)
                    ],
                    source: .online
                )
            }
        }
    }

    // MARK: - Offline Fallback

    private func fallbackToOffline(credential: VerifiableCredential) async -> UnifiedVerificationResult {
        let offlineResult = await offlineVerifier.verifyCredential(credential)

        // Compute bundle age from the cached bundle's fetchedAt timestamp
        let bundleAge: TimeInterval? = await computeBundleAge(credential: credential)

        // Expiry check directly from the credential (not relying on string from offlineResult.error)
        let expiryFailed: Bool = {
            guard let expDateStr = credential.expirationDate,
                  let expDate = ISO8601DateFormatter().date(from: expDateStr) else { return false }
            return expDate < Date()
        }()

        // Build individual checks
        var checks: [VerificationCheck] = []

        // Expiry check
        checks.append(VerificationCheck(
            type: .expiry,
            status: expiryFailed ? .fail : .pass,
            message: expiryFailed ? "Credential has expired" : "Credential is within validity period"
        ))

        // Signature check
        checks.append(VerificationCheck(
            type: .signature,
            status: offlineResult.signatureValid ? .pass : .fail,
            message: offlineResult.signatureValid
                ? "Signature verified using cached bundle"
                : (offlineResult.error ?? "Signature verification failed")
        ))

        // Revocation check
        let revocationCheckStatus: CheckStatus
        let revocationMessage: String
        switch offlineResult.revocationStatus {
        case .valid:
            revocationCheckStatus = .pass
            revocationMessage = "Not revoked (cached status list)"
        case .revoked:
            revocationCheckStatus = .fail
            revocationMessage = "Credential has been revoked"
        case .unknown:
            revocationCheckStatus = .unknown
            revocationMessage = "Revocation status unknown (no cached status list)"
        }
        checks.append(VerificationCheck(
            type: .revocation,
            status: revocationCheckStatus,
            message: revocationMessage
        ))

        // Bundle freshness check
        checks.append(VerificationCheck(
            type: .bundleFreshness,
            status: offlineResult.bundleFresh ? .pass : .fail,
            message: offlineResult.bundleFresh
                ? "Verification bundle is fresh"
                : "Verification bundle is stale or missing"
        ))

        // Determine overall status using the same precedence as Android
        let status = determineStatus(offlineResult: offlineResult, expiryFailed: expiryFailed)

        return UnifiedVerificationResult(
            status: status,
            checks: checks,
            source: .offline,
            bundleAge: bundleAge
        )
    }

    /// Maps OfflineVerificationResult fields to a VerificationStatus,
    /// matching the Android fallback-mapping precedence.
    private func determineStatus(
        offlineResult: OfflineVerificationResult,
        expiryFailed: Bool
    ) -> VerificationStatus {
        if offlineResult.error != nil && !offlineResult.signatureValid {
            // Hard error (missing bundle, serialisation failure, etc.)
            return .failed
        }
        if expiryFailed {
            return .failed
        }
        if !offlineResult.signatureValid {
            return .failed
        }
        if offlineResult.revocationStatus == .revoked {
            return .failed
        }
        if !offlineResult.bundleFresh {
            return .degraded
        }
        if offlineResult.revocationStatus == .unknown {
            return .degraded
        }
        return .verifiedOffline
    }

    private func computeBundleAge(credential: VerifiableCredential) async -> TimeInterval? {
        let issuerDid = Did.fromKeyId(credential.proof.verificationMethod)
        guard let bundle = await bundleStore.getBundle(issuerDid: issuerDid.value),
              let fetchedDate = ISO8601DateFormatter().date(from: bundle.fetchedAt) else {
            return nil
        }
        return Date().timeIntervalSince(fetchedDate)
    }

    // MARK: - Network Error Detection

    private func isNetworkError(_ error: Error) -> Bool {
        let nsError = error as NSError
        if nsError.domain == NSURLErrorDomain { return true }
        if case HttpError.networkError = error { return true }
        if case HttpError.timeout = error { return true }
        if case HttpError.requestFailed(let statusCode, _) = error,
           (500...599).contains(statusCode) { return true }
        return false
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/verifier/offline/VerificationOrchestrator.kt
package my.ssdid.wallet.domain.verifier.offline

import my.ssdid.wallet.domain.model.VerifiableCredential
import my.ssdid.wallet.domain.revocation.RevocationStatus
import my.ssdid.wallet.domain.verifier.Verifier
import java.io.IOException
import java.time.Duration
import java.time.Instant

class VerificationOrchestrator(
    private val onlineVerifier: Verifier,
    private val offlineVerifier: OfflineVerifier,
    private val bundleStore: BundleStore
) {

    suspend fun verify(credential: VerifiableCredential): UnifiedVerificationResult {
        val onlineResult = onlineVerifier.verifyCredential(credential)

        onlineResult.onSuccess { valid ->
            return if (valid) {
                UnifiedVerificationResult(
                    status = VerificationStatus.VERIFIED,
                    checks = listOf(
                        VerificationCheck(CheckType.SIGNATURE, CheckStatus.PASS, "Signature verified online"),
                        VerificationCheck(CheckType.EXPIRY, CheckStatus.PASS, "Credential is not expired"),
                        VerificationCheck(CheckType.REVOCATION, CheckStatus.PASS, "Revocation status confirmed online")
                    ),
                    source = VerificationSource.ONLINE
                )
            } else {
                UnifiedVerificationResult(
                    status = VerificationStatus.FAILED,
                    checks = listOf(
                        VerificationCheck(CheckType.SIGNATURE, CheckStatus.FAIL, "Signature verification failed online")
                    ),
                    source = VerificationSource.ONLINE
                )
            }
        }

        onlineResult.onFailure { error ->
            if (!isNetworkError(error)) {
                return UnifiedVerificationResult(
                    status = VerificationStatus.FAILED,
                    checks = listOf(
                        VerificationCheck(CheckType.SIGNATURE, CheckStatus.FAIL, "Verification error: ${error.message}")
                    ),
                    source = VerificationSource.ONLINE
                )
            }
        }

        // Fall back to offline verification
        return verifyOffline(credential)
    }

    private suspend fun verifyOffline(credential: VerifiableCredential): UnifiedVerificationResult {
        val offlineResult = offlineVerifier.verifyCredential(credential)
        val bundle = bundleStore.getBundle(credential.issuer)
        val bundleAge = bundle?.let {
            try {
                Duration.between(Instant.parse(it.fetchedAt), Instant.now())
            } catch (_: Exception) {
                null
            }
        }

        val expiryFailed = credential.expirationDate?.let {
            try { java.time.Instant.parse(it).isBefore(java.time.Instant.now()) }
            catch (e: Exception) { false }
        } ?: false

        val status = when {
            offlineResult.error != null -> VerificationStatus.FAILED
            expiryFailed -> VerificationStatus.FAILED
            !offlineResult.signatureValid -> VerificationStatus.FAILED
            offlineResult.revocationStatus == RevocationStatus.REVOKED -> VerificationStatus.FAILED
            !offlineResult.bundleFresh -> VerificationStatus.DEGRADED
            offlineResult.revocationStatus == RevocationStatus.UNKNOWN -> VerificationStatus.DEGRADED
            else -> VerificationStatus.VERIFIED_OFFLINE
        }

        val checks = buildList {
            // Signature check
            add(
                VerificationCheck(
                    type = CheckType.SIGNATURE,
                    status = if (offlineResult.signatureValid) CheckStatus.PASS else CheckStatus.FAIL,
                    message = if (offlineResult.signatureValid) "Signature valid (offline)" else "Signature invalid (offline)"
                )
            )

            // Expiry check — uses the expiryFailed computed above
            add(
                VerificationCheck(
                    type = CheckType.EXPIRY,
                    status = if (expiryFailed) CheckStatus.FAIL else CheckStatus.PASS,
                    message = if (expiryFailed) offlineResult.error ?: "Credential expired" else "Not expired"
                )
            )

            // Revocation check
            val revocationCheckStatus = when (offlineResult.revocationStatus) {
                RevocationStatus.VALID -> CheckStatus.PASS
                RevocationStatus.REVOKED -> CheckStatus.FAIL
                RevocationStatus.UNKNOWN -> CheckStatus.UNKNOWN
            }
            add(
                VerificationCheck(
                    type = CheckType.REVOCATION,
                    status = revocationCheckStatus,
                    message = when (offlineResult.revocationStatus) {
                        RevocationStatus.VALID -> "Not revoked"
                        RevocationStatus.REVOKED -> "Credential is revoked"
                        RevocationStatus.UNKNOWN -> "Revocation status unknown"
                    }
                )
            )

            // Bundle freshness check
            add(
                VerificationCheck(
                    type = CheckType.BUNDLE_FRESHNESS,
                    status = if (offlineResult.bundleFresh) CheckStatus.PASS else CheckStatus.FAIL,
                    message = if (offlineResult.bundleFresh) "Bundle is fresh" else "Bundle is stale"
                )
            )
        }

        return UnifiedVerificationResult(
            status = status,
            checks = checks,
            source = VerificationSource.OFFLINE,
            bundleAge = bundleAge
        )
    }

    private fun isNetworkError(error: Throwable): Boolean {
        if (error is IOException) return true
        if (error is retrofit2.HttpException && error.code() in 500..599) return true
        return false
    }
}
